ambry-store/src/main/java/com.github.ambry.store/IndexSegment.java

```diff
@@ -371,6 +380,52 @@ IndexValue find(StoreKey keyToFind) throws StoreException {
     return toReturn;
   }
 
+  /**
+   * Gets values for all matches for {@code keyToFind} at and in the vicinity of {@code positiveMatchInd}.
+   * @param mmap the mmap to read values off of
+   * @param keyToFind the needle {@link StoreKey}
+   * @param positiveMatchInd the index of a confirmed positive match
+   * @param totalEntries the total number of entries in the index
+   * @param values the set to which the {@link IndexValue}s will be added
+   * @return a pair consisting of the lowest index that matched the key and the highest
+   * @throws IOException if there are problems reading from the mmap
+   */
+  private Pair<Integer, Integer> getAllValuesFromMmap(ByteBuffer mmap, StoreKey keyToFind, int positiveMatchInd,
+      int totalEntries, NavigableSet<IndexValue> values) throws IOException {
+    byte[] buf = new byte[valueSize];
+    // add the value at the positive match and anything after that matches
+    int end = positiveMatchInd;
+    StoreKey found;
+    do {
+      mmap.get(buf);
+      values.add(new IndexValue(startOffset.getName(), ByteBuffer.wrap(buf), getVersion()));
+      end++;
+      if (end < totalEntries) {
+        found = getKeyAt(mmap, end);
+        logger.trace("Index Segment {}: found {} at {}", indexFile.getAbsolutePath(), found, end);
+      } else {
+        break;
+      }
+    } while (found.equals(keyToFind));
+    end--;
+
```

Comment:
I think this can be much simpler:
```java
for (end = positiveMatchInd; end < totalEntries && getKeyAt(mmap, end).equals(keyToFind); end++) {
  mmap.get(buf);
  values.add(new IndexValue(startOffset.getName(), ByteBuffer.wrap(buf), getVersion()));
}

for (start = positiveMatchInd - 1; start >= 0 && getKeyAt(mmap, start).equals(keyToFind); start--) {
  mmap.get(buf);
  values.add(new IndexValue(startOffset.getName(), ByteBuffer.wrap(buf), getVersion()));
}
```

Reply:
Ah - this is great. Thanks !